Migrate block email editor to the wordpress/editor Editor component

costasovo opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

This PR is a step towards migrating the block email editor to WordPress's Block Editor. In this PR, we migrate from the custom email editor to the Editor component, which is used in the site editor and the post editor, and covers a lot of functionality we replicated (e.g., whole UI around canvas) or missed in the custom editor implementation.

We do this migration to improve the alignment with core's editors and to reduce the codebase we need to maintain and update to keep up with the changes in the core.

Closes WOOPLUG-3951

### Screenshots or screen recordings:

<!-- If this PR includes UI changes, please provide screenshots or a screen recording for clarity. -->
<!-- This section can be removed if not applicable. -->

| Before | After |
| ------ | ----- |
|    <img width="1522" alt="Screenshot 2025-05-06 at 17 17 42" src="https://github.com/user-attachments/assets/cc293576-7ab2-4d88-b9a9-084f5adae038" />  |  <img width="1521" alt="Screenshot 2025-05-07 at 9 21 46" src="https://github.com/user-attachments/assets/369088ff-bfa2-449f-bd4a-fc305f4cd885" /> |


<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

#### Note for reviewer:
Sorry for the big PR. I tried my best to split changes into small commits and explain changes in the commit description.

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Enable the Block Email Editor in Woo > Settings > Advanced > Features
2. Go to Woo > Settings > Emails and try editing emails. 
3. Test also the Edit template mode (Woo > Settings > Emails > Edit template)
4. To verify that we can create a new email with the Select template modal, you need to enable native listing for Woo Email CPT
add following to
```php
'public'       => false,
'show_ui'      => true,
'show_in_menu' => true,
'show_in_rest' => true,
```
to 

https://github.com/woocommerce/woocommerce/blob/1817abf09affb7e33fb0c4d68cdecf6a810536c4/plugins/woocommerce/src/Internal/EmailEditor/Integration.php#L122-L139			

Overall, please try to compare it with the version before the refactor and look for regressions.

<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

costasovo committed:
Add private APIs for the Editor component

costasovo committed:
Replace the custom editor layout with the Editor component

costasovo committed:
Move layout styles into the useEmailCss hook

The logic for restricting email width to the width defined in the layout settings
was in the VisualEditor component.

The VisualEditor is not used any more so I moved the logic to the hook,
that generates CSS for the editor content.

WOOPLUG-3951

costasovo committed:
Fix PageRenderer to support creating new emails

Post ID might be 0 when we create a new email so we must use
Post ID from the newly created email post.
WOOPLUG-3951

costasovo committed:
Reintroduce TemplateSelection functionality

WOOPLUG-3951

costasovo committed:
Remove unnecessary patterns fetching

The patterns are fetched within the Editor component

WOOPLUG-3951

costasovo committed:
Migrate email styles sidebar to PluginSidebar API

WOOPLUG-3951

costasovo committed:
Add custom header to the email styles sidebar

The default PluginSidebar header contains a button with a star icon that
completely hides the sidebar even from the top editor bar.
We don't have such functionality in designs.

WOOPLUG-3951

costasovo committed:
Update Navigator component in the email styles sidebar

Navigator is now non-experimental  component.
https://wordpress.github.io/gutenberg/?path=/docs/components-navigator--docs
WOOPLUG-3951

costasovo committed:
Update Woo Email CPT labels

These labels are displayed in some places of the editor (e.g. Email
in the sidebar tabs). Woo Email is the internal name.

WOOPLUG-3951

costasovo committed:
Add send preview to preview options

PluginPreviewMenuItem adds a new item in the preview toggle dropdown.
WOOPLUG-3951

costasovo committed:
Re-enable fullscreen mode

The full-screen mode controls are not part of wordpress/editor.
We need to add more menu via ViewMoreMenuGroup component and control
its' state in our store.

Inspired by edit-post package.

WOOPLUG-3951

costasovo committed:
Add autosave, error boundary, basic shortcuts, and snackbar notices

WOOPLUG-3951

costasovo committed:
Hide original status panel

The original status pabel contains some options that don't work
for emails (e.g., user can switch to a site template), and we can't hide it selectively.

WOOPLUG-3951

costasovo committed:
Add settings panel for emails

WOOPLUG-3951

costasovo committed:
Add sender settings to template mode

We currently can't use PluginPostStatusInfo slot because post-status
panel is hidden.
We add the sender settings as an extra panel.
WOOPLUG-3951

costasovo committed:
Introduce Command menu with basic commands

WOOPLUG-3951

costasovo committed:
Remove dependency on @wordpress/interface

These actions are no longer needed as they are completely covered
by the Editor component.

WOOPLUG-3951

costasovo committed:
Fix fetching an email not associated with transactional email

When $email_type was null it caused 500 because get_email_by_type expects string.
WOOPLUG-3951

costasovo committed:
Add the Save Email button

I placed the code intentionally in the hack folder.
It is one of the things we should address upstream in the future.
Currently, we hide the original button when there is nothing
to save and we replace it with our custom button.

WOOPLUG-3951

costasovo committed:
Configure autosave timeout and disable code editor

WOOPLUG-3951

costasovo committed:
Add validation notices and custom snackbars

The custom snackbars are included in EditorNotices and contain
text replacement functionality to ensure notices are understandable in email context.

I added additional hacky slot for rendering validation notice into
the proper area.

WOOPLUG-3951

costasovo committed:
Fix tiny CSS issues

- floating preview text lenght info was not counted to element height
- slightly missplaced icon in preview dropdown

costasovo committed:
Fix global padding is not applied

This was previously working because se manually added
class 'has-global-padding' to the root container.
We no longer have access to add class to the container so
we need to generate CSS ourselves.

WOOPLUG-3951

costasovo committed:
Fix loading CSS from the server into the editor

This was previously done in a low-level component we no longer use.

WOOPLUG-3951

costasovo committed:
Remove the unwanted margin for the first content element

The editor has newly class wp-site-blocks added automatically and this rule
was conflicting how we set margins for block gap settings.

I also removed the comment because it was not a valid CSS comment.
WOOPLUG-3951

costasovo committed:
Hide the huge title input

In the preview dropdown, there is an option to hide the template content.
In such case, the editor switches to post-only mode and displays the big
title input on top. Such a title makes no sense in an email context as it
is not part of the final render.

This commit adds CSS to hide it.
WOOPLUG-3951

costasovo committed:
Fix useSelect warning about returning different values for template

The `useSelect` hook returns different values when called with the same state and parameters.
This can lead to unnecessary re-renders and performance issues if not fixed.

Non-equal value keys: template

WOOPLUG-3951

costasovo committed:
Fix the useSelect warning about globalStylesPost

There was a strange fallback to template post when global styles post was not present.
After I removed the fallback I had to make a few fixes in style hooks.

Warning was:

The `useSelect` hook returns different values when called with the same state and parameters.
This can lead to unnecessary re-renders and performance issues if not fixed.

Non-equal value keys: globalStylesPost

WOOPLUG-3951

costasovo committed:
Cleanup unused components from the email editor

These components are no longer used after we switched to the Editor
WOOPLUG-3951

costasovo committed:
Cleanup SCSS files and fix CSS loading

The CSS loading issue was caused by content related CSS for richtext defined in CSS which is used for layout
and was added additionally to the iframe.

I copied styles that we need for richtext to content-editor.css, which
is designated for CSS that we want to apply to content.

WOOPLUG-3951

costasovo committed:
Cleanup private APIs

WOOPLUG-3951

costasovo committed:
Fix JS lint issues

WOOPLUG-3951

costasovo committed:
Cleanup editore store

WOOPLUG-3951

costasovo committed:
Add changelogs

WOOPLUG-3951

costasovo committed:
Optimize fetching of emailPosts

We need email posts only for the template select modal for the new email.
When we fetch previews only for swapping templates, we can fetch sent emails.

WOOPLUG-3951

triple0t reviewed (commented):
Incredible work. Well done, @costasovo 

Feedback from my first look at the UI. I haven't dived into the codebase and commits yet.

1. The Preferences button is not working. Clicking on it does not do anything
![Screenshot 2025-05-08 at 12 40 44 PM](https://github.com/user-attachments/assets/cc27e429-526a-4734-bda7-65d9a25b74da)
   * We don't have it on the old editor. So maybe fix it, or we hide/remove it.

2. Border: We only need to support `px`. The value can currently be saved in px, %, em, rem, etc
![Screenshot 2025-05-08 at 12 46 26 PM](https://github.com/user-attachments/assets/a90522b9-a161-4538-b054-bd9a2f57d08f)
![Screenshot 2025-05-08 at 12 45 33 PM](https://github.com/user-attachments/assets/7b29c896-40da-47a5-8999-ff32cc5ec70b)
![Screenshot 2025-05-08 at 12 53 49 PM](https://github.com/user-attachments/assets/46adc59d-6da2-4f0e-adde-3cc593988b4a)


3. Colors: we only need to support `hex`.
![Screenshot 2025-05-08 at 12 47 21 PM](https://github.com/user-attachments/assets/b2b64995-8d41-4576-aa37-7b4cb0d3bdeb)


4. Saving email shows the post updated in the Snackbar.
![Screenshot 2025-05-08 at 12 48 26 PM](https://github.com/user-attachments/assets/a49cfefe-3e5a-40b4-a6c8-fe8051432d79)
![Screenshot 2025-05-08 at 12 48 42 PM](https://github.com/user-attachments/assets/ec1e8f27-048a-4803-8313-92f80ac788e5)


5. Why "Send preview", not "Send test email" or "Send email preview"?
![Screenshot 2025-05-08 at 2 32 28 PM](https://github.com/user-attachments/assets/af57d316-cab2-4448-89cc-bdd61f0de893)


6. There's a problem with the ListView: ![ListView.gif](https://github.com/user-attachments/assets/82f1af42-149b-4e0e-8697-85e6488f1ba3) 
Comparing: ![9SRelj.gif](https://github.com/user-attachments/assets/38982bf0-0022-4b89-b9ae-aa238d6660c3)

7. Do we want to allow folks to edit the Email Template description?
![Screenshot 2025-05-08 at 1 25 15 PM](https://github.com/user-attachments/assets/77928ecd-536c-4a10-8f59-88fc27cbd191)


8. ~Saving an email does not redirect back to the email listing page~ I needed to click the button twice. 
Save email button needing to be clicked twice: What if we add support for this https://developer.wordpress.org/block-editor/reference-guides/slotfills/main-dashboard-button/ in which we can add an icon to close or return back to a page?


**What went well.**
I can see all available blocks
![Screenshot 2025-05-08 at 1 17 31 PM](https://github.com/user-attachments/assets/051b2447-fc6a-4cc0-81af-d3158430e4f8)

The Interface works as expected
![Screenshot 2025-05-08 at 1 37 22 PM](https://github.com/user-attachments/assets/3b565c6a-8c85-4748-a92c-3e9537e72764)

The new email template selector works
![Screenshot 2025-05-08 at 2 36 31 PM](https://github.com/user-attachments/assets/19230bfd-0b0c-4424-a13d-384a261d81fb)

triple0t commented on packages/php/email-editor/src/Engine/content-editor.css:
```diff
@@ -114,3 +114,41 @@ ol.has-background {
+ * Hide the post title.
+ * When user disables the template-lock mode we don't want to show the post title.
+ */
+.editor-visual-editor__post-title-wrapper {
+	display: none;
+}
```
Can you share a screenshot of what the style is fixing?
When I disabled the style, I didn't notice any difference.

costasovo replied:
The CSS is effective when user disables "Show template" in the preview dropdown.

![Screenshot 2025-05-12 at 16 39 08](https://github.com/user-attachments/assets/4963b7bb-3ac2-48f2-85cd-8afa8b93d8c8)

Without the CSS, when the "Show template" is disabled, there is a huge editable email Title displayed above the email content.

![Screenshot 2025-05-12 at 16 43 50](https://github.com/user-attachments/assets/816ea643-3b68-4dd3-a517-11b48295558e)


We can't easily remove the "Show template" toggle from the preview options and IMO it may make sense for some users who want to focus purely on content also helps to educate that there is a template + content concept so I think it is ok to keep it. I hide the title from the canvas because, unlike a web post we don't render the title in the email content.

triple0t commented on packages/php/email-editor/src/Engine/content-editor.css:
```diff
@@ -63,7 +63,7 @@
  */
 .wp-block {
-	clear: both; // for ensuring that floated elements (images) are cleared
+	clear: both;
```
Thanks for fixing this

triple0t commented on packages/js/email-editor/src/hooks/use-email-styles.ts:
```diff
@@ -150,11 +150,11 @@ export const useEmailStyles = (): EmailStylesData => {
+	const styles = useMemo( () => {
+		return userTheme
+			? cleanupUserStyles( shortenWpPresetVariables( userTheme?.styles ) )
+			: {};
```
To prevent a lot of unnecessary renders, I think we should provide a constant Empty Object or Empty Array for use in components.

We did something similar [here](https://github.com/woocommerce/woocommerce/blob/038bfbee7e16b98903c0634af70af60fdfb74c05/packages/js/email-editor/src/hooks/use-content-validation.ts#L19-L21), and [here](https://github.com/woocommerce/woocommerce/blob/99f04562c084808e02515a526664fdc0c77fa688/plugins/woocommerce/client/blocks/assets/js/data/collections/selectors.js#L10) is an example use in Woo Blocks.

costasovo replied:
Implemented in https://github.com/woocommerce/woocommerce/pull/57775/commits/ad3d3342c05437da97d65cb6c4f77df091741125

triple0t commented on packages/js/email-editor/src/hacks/publish-save.tsx:
```diff
@@ -0,0 +1,111 @@
+};
+
+function NextPublishSlot( { children }: NextButtonSlotPropType ) {
+	const [ sendButtonPortalEl ] = useState( document.createElement( 'div' ) );
```
The decision to use `useState` here instead of `useRef` to store the Node Element is quite interesting. May I ask why?

costasovo replied:
Oh! Nice catch! No reason. It is a mistake or a typo, probably created when I was exploring possible solutions. It seemed to work because we actually never changed that state. Updated in https://github.com/woocommerce/woocommerce/pull/57775/commits/ad3d3342c05437da97d65cb6c4f77df091741125

triple0t commented on packages/js/email-editor/src/hacks/publish-save.tsx:
```diff
@@ -0,0 +1,111 @@
+export function PublishSave() {
+	const { validateContent, isInvalid } = useContentValidation();
+	const observerRef = useRef< MutationObserver | null >( null );
+	const { dirtyEntityRecords } = useEntitiesSavedStatesIsDirty();
```
This is pretty incredible, but looking at it from a performance point of view, it is difficult to accept.
The component will render and re-render for every unrelated change in the editor canvas because of  `useEntitiesSavedStatesIsDirty`, `useContentValidation`, and `MutationObserver`

Since we need to replace the `Save` button when the post is saved, why not subscribe only to that notification?

Options
1. Listen for doAction [hook](https://github.com/WordPress/gutenberg/blob/bc28417b3a95c1ab8fb8f5c67a98dbf31a6cf3ec/packages/editor/src/store/actions.js#L279-L283)
2. Other selectors: didPostSaveRequestSucceed
3. If none of that works, we could use a [custom](https://github.com/WordPress/gutenberg/blob/7e13df08782341db799138e943b56d037cd7fe4e/packages/editor/src/components/editor-interface/index.js#L51) save button. Skipping the whole hack and replacement stuff.

costasovo replied:
Good catch! Unfortunately, we can't use Option 1 or 2, because we need to change the button also when the user types. I wanted to avoid the custom button, because we would need to implement much more.
In the end, I used two boolean selectors from the editor store and moved the validation hook to a subcomponent.
So, currently, there are no re-renders when typing in Canvas. Done in https://github.com/woocommerce/woocommerce/pull/57775/commits/ad3d3342c05437da97d65cb6c4f77df091741125

triple0t commented on packages/js/email-editor/src/hacks/publish-save.tsx:
```diff
@@ -0,0 +1,111 @@
+		const publishButton = document.getElementsByClassName(
+			'editor-post-publish-button__button'
+		)[ 0 ];
+		toggleElementVisible( publishButton, displayOriginalPublishButton );
+
+		// It may get additionally re-rendered by the editor, so we need to check it again
+		if ( observerRef.current ) {
+			observerRef.current.disconnect();
+		}
+		observerRef.current = new MutationObserver( () => {
+			toggleElementVisible( publishButton, displayOriginalPublishButton );
+		} );
+		observerRef.current.observe( publishButton, {
+			attributes: true,
+			childList: true,
+			subtree: false,
+		} );
```
Let's wrap this block of code in an if statement.
We're going to have a bunch of errors here if the Node element can not be found.

costasovo replied:
Done in https://github.com/woocommerce/woocommerce/pull/57775/commits/ad3d3342c05437da97d65cb6c4f77df091741125

triple0t commented on packages/js/email-editor/src/components/styles-sidebar/styles-sidebar.tsx:
```diff
@@ -3,19 +3,15 @@
-	__experimentalNavigatorProvider as NavigatorProvider,
-	__experimentalNavigatorScreen as NavigatorScreen,
-} from '@wordpress/components';
+import { Navigator } from '@wordpress/components';
```
Thank you for migrating to the stable component.

triple0t commented on packages/js/email-editor/src/components/sidebar/template-settings-panel.tsx:
```diff
@@ -20,10 +22,14 @@ export function TemplateSettingsPanel() {
 	return (
-		<>
+		<PluginDocumentSettingPanel
+			name="template-settings-panel"
+			title={ __( 'Settings', 'woocommerce' ) }
+			className="woocommerce-email-editor__settings-panel"
+		>
 			{ templateSections.map( ( section ) => (
 				<div key={ section.id }>{ section.render() }</div>
 			) ) }
-		</>
+		</PluginDocumentSettingPanel>
```
I think at this point, we should wrap this component in a custom error boundary so that errors that occur from third-party integration do not totally break the whole email editor interface.

costasovo replied:
Added in https://github.com/woocommerce/woocommerce/pull/57775/commits/a47a9728b23a30e19f8819d40ae55d0ec21f7435

triple0t commented on packages/js/email-editor/src/components/sidebar/style.scss:
```diff
@@ -8,17 +8,3 @@
-	}
-
-	// Because we use complementary area component from the WP interface package the is visible an unwanted close icon.
-	// This style hides it.
-	.interface-complementary-area-header__small {
-		display: none;
-	}
-}
```
We can remove this as well. https://github.com/woocommerce/woocommerce/blob/173bae563b0e0741f3c1cb63760696a7715be2a1/packages/js/email-editor/src/components/styles-sidebar/style.scss#L55-L59

triple0t commented on packages/js/email-editor/src/store/actions.ts:
```diff
@@ -1,139 +1,31 @@
  * External dependencies
  */
-import { dispatch, select } from '@wordpress/data';
-import { store as interfaceStore } from '@wordpress/interface';
```
Since we don't need `@wordpress/interface` anymore, can we remove these
* https://github.com/woocommerce/woocommerce/blob/173bae563b0e0741f3c1cb63760696a7715be2a1/packages/js/email-editor/src/types/index.ts#L7
* https://github.com/woocommerce/woocommerce/blob/173bae563b0e0741f3c1cb63760696a7715be2a1/packages/js/email-editor/src/types/index.ts#L110
* https://github.com/woocommerce/woocommerce/blob/173bae563b0e0741f3c1cb63760696a7715be2a1/packages/js/email-editor/src/types/wordpress-modules.ts#L48-L67

triple0t reviewed (commented):
Well done, @costasovo 

Thanks for implementing this.

Some notes

1. The Global styles are not loading (and not applied) on the Editor for non-admin users. Preview works fine. This works with the former Editor.
2. We need to fix type errors `pnpm --filter='@woocommerce/email-editor' check-types`
3. Question: Since we are using the Editor now, do we still need the `SlotFillProvider` component?

costasovo committed:
Hide "Preferences" item in the more menu

Currently we don't have any preferences to display.
When we fully migrate to the post editor we need to decide
whether we use the preferences from the post editor or keep
the memu item hidden.

WOOPLUG-3951

costasovo committed:
Update the snackbar notice when updating the email post.

WOOPLUG-3951

costasovo committed:
Update preview dropdown item text for sending preview

The new text is the same as the one used in the version before the migration.
WOOPLUG-3951

costasovo committed:
Improve memoization in use-email-styles hook

WOOPLUG-3951

costasovo committed:
Fixes and improvements in the publish-save hack

- fixes re-render issues by using proper selectors and moving the validation hook to SendButton component
- replaces incorrectly used useState with useRef to hold a reference of a DOM node
- add check if publish button node is present before performing any changes on that

WOOPLUG-3951

costasovo committed:
Add error boundary for extenders of woocommerce_email_editor_template_sections filter

I chose to add the boundary per section so that the only the problematic sections
so that we can still render sections that work.
WOOPLUG-3951

costasovo committed:
Cleanup types enhacements for wordpress/interface

WOOPLUG-3951

costasovo committed:
Fix the style loading for the editor user

WOOPLUG-3951

costasovo committed:
Address types and linter errors

WOOPLUG-3951

costasovo committed:
Reintroduce block compatibility warnings in the sidebar

WWOOPLUG-3951

costasovo commented:
@triple0t Thank you for the thorough review! Much appreciated.

> 1. The Preferences button is not working. Clicking on it does not do anything

Fixed in https://github.com/woocommerce/woocommerce/pull/57775/commits/334ca2f0546ff9993fd24610920e64c4cbf60b3b

> 2. Border: We only need to support px. The value can currently be saved in px, %, em, rem, etc

This problem is not new. It is already present in the current trunk. I created an extra issue for that https://github.com/woocommerce/woocommerce/issues/57957

> 3. Colors: we only need to support hex.

I found that even when a user uses RGB to configure the color we render hex in the renderer so IMO it doesn't need fixing. Only problem is opacity, but we already have a warning saying "Border display may vary or be unsupported in some email clients." so it is IMO covered.

> 4. Saving email shows the post updated in the Snackbar.

Addressed in https://github.com/woocommerce/woocommerce/pull/57775/commits/7d6fc4b76b9f38c650bc8e3a6802bc703fb8dd71

> 5. Why "Send preview", not "Send test email" or "Send email preview"?

Good catch. It was kind of random text. Updated in https://github.com/woocommerce/woocommerce/pull/57775/commits/7d6fc4b76b9f38c650bc8e3a6802bc703fb8dd71

> 6 There's a problem with the ListView:

I found the lags in opening ListView when a block is selected are related to `template-lock` mode. I was able to replicate that also in the core's post editor. I didn't pursue that further. 

> 7 Do we want to allow folks to edit the Email Template description?

It is a feature we get "Out of the box". It is not helpful for Woo Integration, as we have only one template. But it also does no harm. In general, I think it is useful as users might want to add some additional info for their templates. 

> 8 I needed to click the button twice.
Save email button needing to be clicked twice: What if we add support for this https://developer.wordpress.org/block-editor/reference-guides/slotfills/main-dashboard-button/, in which we can add an icon to close or return to a page?

Currently, you still need to click Save and then Save email if there are some changes. This will be addressed in https://github.com/woocommerce/woocommerce/issues/57959

> The Global styles are not loading (and not applied) on the Editor for non-admin users. Preview works fine. This works with the former Editor.

Fixed in https://github.com/woocommerce/woocommerce/pull/57775/commits/c97f87cbe7a37cfcf8250fec89c999e8e8555c8d

> We need to fix type errors pnpm --filter='@woocommerce/email-editor' check-types

Fixed in https://github.com/woocommerce/woocommerce/pull/57775/commits/a0b30102fc977b8e907aa37a670c1a464dd09aa3

Note: I fixed most of the warnings by adding ts-expect-error. I think we need to update our local WP packages to something newer than the current wp-6.6. I tried to do that, but it seems to require more time, so I'll prepare a ticket for that.

> Question: Since we are using the Editor now, do we still need the SlotFillProvider component?

I actually copied that from [edit-post's Layout](https://github.com/WordPress/gutenberg/blob/918d152f25248a53e3fedd1ec3af2eadeb038e9d/packages/edit-post/src/components/layout/index.js#L584). I think it is needed for modals and command menu.

I hope I didn't miss any of your remarks. I also added a few small fixes. New commits start from https://github.com/woocommerce/woocommerce/pull/57775/commits/334ca2f0546ff9993fd24610920e64c4cbf60b3b

triple0t reviewed (approved):
Looks good to me. Thank you for implementing the change.
Please feel free to merge when ready.
